packages/minter-contracts/test-hs/Test/EnglishAuction/FA2FixedFee.hs

```diff
@@ -0,0 +1,744 @@
+  , testMaxConfigToStartTime :: Natural
+  , testTimeToStart :: Integer
+  , testAuctionDuration :: Natural
+  , testMaxAuctionTime :: Natural
+
+  , testOpeningPrice :: Natural
+  , testFirstBid :: Natural
+  , testMinRaise :: Natural
+  , testMinRaisePercent :: Natural
+
+  , testTokenBatches :: [[FA2Token]]
+
+  , testBidTokenId :: TokenId
+  }
+  deriving stock (Show)
+
+genTestData :: Gen TestData
+genTestData = do
+  testFeePercent <- Gen.integral (Range.linear 0 100)
+  testRoundTime <- Gen.integral (Range.linear 1 100)
+  testExtendTime <- Gen.integral (Range.linear 1 100)
+
+  testMaxConfigToStartTime <- Gen.integral (Range.linear 0 100)
+  testTimeToStart <- Gen.integral (Range.linear 0 (toInteger testMaxConfigToStartTime))
+
+  testOpeningPrice <- Gen.integral (Range.linear 1 1000)
+  testFirstBid <- Gen.integral (Range.linear testOpeningPrice (testOpeningPrice + 1000))
+  testMinRaise <- Gen.integral (Range.linear 0 1000)
+  testMinRaisePercent <- Gen.integral (Range.linear 0 100)
+
+  testAuctionDuration <- Gen.integral (Range.linear 1 100)
+  testMaxAuctionTime <- Gen.integral (Range.linear testAuctionDuration (testAuctionDuration * 2))
+
+  testTokenBatches <-
+    Gen.list (Range.linear 0 10) $
+        Gen.list (Range.linear 0 10) $
+        FA2Token
+          <$> (TokenId <$> Gen.integral (Range.linear 0 10))
+          <*> Gen.integral (Range.linear 0 10)
+
+  testBidTokenId <- TokenId <$> Gen.integral (Range.linear 0 10)
+
+  pure $ TestData {..}
+
+-- | Generates a list with 0 or more valid bids
+genSomeBids :: TestData -> Gen (NonEmpty Natural)
+genSomeBids testData = do
+  len <- Gen.int (Range.linear 0 5)
+  firstBid <- genBid testData (testOpeningPrice testData)
+  moreBids <- iterateM len (genBid testData) firstBid
+  pure $ firstBid :| moreBids
+
+-- | Generates a list with 1 or more valid bids
+genManyBids :: TestData -> Gen [Natural]
+genManyBids testData = do
+  len <- Gen.int (Range.linear 0 5)
+  iterateM len (genBid testData) (testOpeningPrice testData)
+
+-- | Generate a valid bid, such that it is:
+--
+-- * Greater than the previous bid or opening price.
+-- * Above either the `min_raise` or the `min_raise_percent` thresholds.
+genBid :: TestData -> Natural -> Gen Natural
+genBid testData@TestData{testMinRaise} previousBid = do
+  raiseLowerBound <- Gen.element [testMinRaise, minRaisePercentOfBid testData previousBid]
+  raise <- Gen.integral (Range.linear raiseLowerBound (raiseLowerBound + 1000))
+  pure $ previousBid + raise
+
+-- | Given a generator of values of type @a@ and an initial value,
+-- repeatedly uses the generator to create a list of the given length,
+-- feeding it the previously generated value at each iteration.
+iterateM :: forall a. Int -> (a -> Gen a) -> a -> Gen [a]
+iterateM len gen initial = do
+  go len initial
+  where
+    go :: Int -> a -> Gen [a]
+    go 0 _ = pure []
+    go n previous = do
+      current <- gen previous
+      (current :) <$> go (n - 1) current
```

Comment:
Looks like the inner function is unnecessary:
```suggestion
iterateM :: forall a. Int -> (a -> Gen a) -> a -> Gen [a]
iterateM 0 _ _ = pure []
iterateM n gen previous = do
  current <- gen previous
  (current :) <$> iterateM (n - 1) current
```

Reply:
Good catch :D When I initially wrote it, `iterateM` was also generating the length of the list, but then I moved this generator out of `iterateM`.